persistence/pre_persistence/persistence.go

```diff
@@ -0,0 +1,306 @@
+		err := newDB.Put(KeyFromHeightKey(it.Key()), it.Value())
+		if err != nil {
+			return nil, err
+		}
+	}
+	context := &PrePersistenceContext{
+		Height: height,
+		Parent: m,
+		DBs:    make([]*memdb.DB, 0),
+	}
+	context.DBs = append(context.DBs, newDB)
+	return context, nil
+}
+
+func (m *PrePersistenceModule) GetCommitDB() *memdb.DB {
+	return m.CommitDB
+}
+
+type PrePersistenceContext struct {
+	Height     int64
+	Parent     modules.PersistenceModule
+	SavePoints map[string]int // TODO save points not entirely implemented. Happy path only for now, rollbacks for later
+	DBs        []*memdb.DB
+}
+
+func (m *PrePersistenceContext) GetLatestBlockHeight() (uint64, error) {
+	return uint64(m.Height), nil
+}
+
+// ExportState Unused but high potential for usefulness for telemetry
+func (m *PrePersistenceContext) ExportState() (*GenesisState, types.Error) {
+	var err error
+	state := &GenesisState{}
+	state.Validators, err = m.GetAllValidators(m.Height)
+	if err != nil {
+		return nil, types.ErrGetAllValidators(err)
+	}
+	state.Apps, err = m.GetAllApps(m.Height)
+	if err != nil {
+		return nil, types.ErrGetAllApps(err)
+	}
+	state.Fishermen, err = m.GetAllFishermen(m.Height)
+	if err != nil {
+		return nil, types.ErrGetAllFishermen(err)
+	}
+	state.ServiceNodes, err = m.GetAllServiceNodes(m.Height)
+	if err != nil {
+		return nil, types.ErrGetAllServiceNodes(err)
+	}
+	state.Pools, err = m.GetAllPools(m.Height)
+	if err != nil {
+		return nil, types.ErrGetAllPools(err)
+	}
+	state.Accounts, err = m.GetAllAccounts(m.Height)
+	if err != nil {
+		return nil, types.ErrGetAllAccounts(err)
+	}
+	state.Params, err = m.GetParams(m.Height)
+	if err != nil {
+		return nil, types.ErrGetAllParams(err)
+	}
+	return state, nil
+}
+
+// NewSavePoint Create a save point
+// Needed for atomic rollbacks in the case of failed transactions during proposal or blocks during validation
+func (m *PrePersistenceContext) NewSavePoint(bytes []byte) error {
+	index := len(m.DBs)
+	newDB := NewMemDB()
+	if index == 0 {
+		return fmt.Errorf("%s", "zero length mock persistence context")
+	}
+	src := m.DBs[index-1]
+	if err := CopyMemDB(src, newDB); err != nil {
+		return err
+	}
+	m.SavePoints[hex.EncodeToString(bytes)] = index
+	m.DBs = append(m.DBs, newDB)
+	return nil
+}
+
+// RollbackToSavePoint Rollback save point
+// Needed in the case of failed transactions during proposal or blocks during validation
+func (m *PrePersistenceContext) RollbackToSavePoint(bytes []byte) error {
+	rollbackIndex, ok := m.SavePoints[hex.EncodeToString(bytes)]
+	if !ok {
+		return fmt.Errorf("save point not found")
+	}
+	toDelete := make([]string, 0)
+	// rollback savepoints map
+	for key, i := range m.SavePoints {
+		if i > rollbackIndex {
+			toDelete = append(toDelete, key)
+		}
+	}
+	for _, key := range toDelete {
+		delete(m.SavePoints, key)
+	}
+	// rollback
+	m.DBs = m.DBs[:rollbackIndex]
+	return nil
+}
+
+// AppHash creates a unique hash based on the global state object
```

Comment:
Is `based on the global state` a requirement or a measure to guarantee maximum entropy?

Reply:
Yes